diag_manager/fms_diag_field_object.F90

```diff
@@ -359,7 +368,115 @@ subroutine set_vartype(objin , var)
           " r8, r4, i8, i4, or string.", warning)
  end select
 end subroutine set_vartype
-
+!> Allocates the data buffer in the field object.
+!! Adds the input data to the buffered data.
+subroutine set_data_buffer (this, input_data, diag_axis)
+  class (fmsDiagField_type) , intent(inout):: this !< The field object
+  class(*), dimension(:,:,:,:), intent(in) :: input_data !< The input array
+  class(fmsDiagAxisContainer_type),intent(in)   :: diag_axis(:)          !< Array of diag_axis
+!> Allocate the buffer if it is not allocated
+  if (.not.allocated(this%data_buffer_allocated)) this%data_buffer_allocated = .false.
+  if (.not.this%data_buffer_allocated) &
+    this%data_buffer_allocated =  allocate_data_buffer(this, input_data, diag_axis)
+  if (.not.this%data_buffer_allocated) &
+    call mpp_error ("set_data_buffer", "The data buffer for the field "//trim(this%varname)//" was unable to be "//&
+      "allocated.", FATAL)
+
+!> Buffer a copy of the data
+  select type (input_data)
+    type is (real(kind=r4_kind))
+      select type (db => this%data_buffer)
+        type is (real(kind=r4_kind))
+          db = input_data
+      end select
+    type is (real(kind=r8_kind))
+      select type (db => this%data_buffer)
+        type is (real(kind=r8_kind))
+          db = input_data
+      end select
+    type is (integer(kind=i4_kind))
+      select type (db => this%data_buffer)
+        type is (integer(kind=i4_kind))
+          db = input_data
+      end select
+    type is (integer(kind=i8_kind))
+      select type (db => this%data_buffer)
+        type is (integer(kind=i8_kind))
+          db = input_data
+      end select
+    class default
+        call mpp_error ("set_data_buffer", "The data input to set_data_buffer for "//&
+          trim(this%varname)//" does not match the buffer for the field object",  FATAL)
+  end select
+end subroutine set_data_buffer
+!> Allocates the global data buffer for a given field using a single thread. Returns true when the
+!! buffer is allocated
+logical function allocate_data_buffer(this, input_data, diag_axis)
+  class (fmsDiagField_type), target, intent(inout):: this !< The field object
+  class(*), dimension(:,:,:,:), intent(in) :: input_data !< The input array
+  class(fmsDiagAxisContainer_type),intent(in)   :: diag_axis(:) !< Array of diag_axis
+  integer :: naxes !< The number of axes in the field
+  integer, parameter :: ndims = 4
```

Comment:
what about the diurnal axis? And i think the math functions are going to assume that the diurnal axis is the fourth dimension?

Reply:
Actually this is the data is being sent in `send_data`, so the diurnal axis does not matter here.